src/stores/AppStore.ts

```diff
@@ -929,30 +931,39 @@ export class AppStore {
         try {
             const ack = await this.backendService.importRegion(directory, file, type, frame.frameInfo.fileId);
             if (frame && ack.success && ack.regions) {
-                const regionMap = new Map<string, CARTA.IRegionInfo>(Object.entries(ack.regions));
-                const regionStyles = new Map<string, CARTA.IRegionStyle>(Object.entries(ack.regionStyles));
-                regionMap.forEach((regionInfo, regionIdString) => {
-                    const styleInfo = regionStyles.get(regionIdString);
-
-                    frame.regionSet.addExistingRegion(
-                        regionInfo.controlPoints as Point2D[],
-                        regionInfo.rotation,
-                        regionInfo.regionType,
-                        parseInt(regionIdString),
-                        styleInfo?.name,
-                        styleInfo?.color,
-                        styleInfo?.lineWidth,
-                        styleInfo?.dashList
-                    );
-                });
+                const regions = Object.entries(ack.regions);
+                const regionStyleMap = new Map<string, CARTA.IRegionStyle>(Object.entries(ack.regionStyles));
+                this.addRegionsInBatch(regions, regionStyleMap, 0);
             }
-            this.fileBrowserStore.hideFileBrowser();
         } catch (err) {
             console.error(err);
             AppToaster.show(ErrorToast(err));
         }
     };
 
+    addRegionsInBatch = (regions: [string, CARTA.IRegionInfo][], regionStyleMap: Map<string, CARTA.IRegionStyle>, count: number) => {
```

Comment:
shouldn't this be an `@action` in order to batch side-effects of adding each region?

Reply:
it does not modify any observable directly, so that mobx did not give any warning.